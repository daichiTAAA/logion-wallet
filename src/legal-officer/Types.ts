import { ColorTheme, rgbaToHex } from '../component/ColorTheme';
import Identity from '../component/types/Identity';
import PostalAddress from '../component/types/PostalAddress';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

const LEGAL_OFFICERS: string[] = [ DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER ];

export function isLegalOfficer(address: string): boolean {
    return LEGAL_OFFICERS.includes(address);
}

export type TokenizationRequestStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface AssetDescription {
    assetId: string,
    decimals: number,
}

export interface TokenizationRequest {
    id: string,
    legalOfficerAddress: string,
    requesterAddress: string,
    requestedTokenName: string,
    bars: number,
    status: TokenizationRequestStatus,
    rejectReason?: string | null,
    createdOn?: string,
    decisionOn?: string,
    assetDescription?: AssetDescription,
}

export type LegalOfficerDecisionStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface LegalOfficerDecision {
    legalOfficerAddress: string,
    status: LegalOfficerDecisionStatus,
    rejectReason: string | null,
    decisionOn: string | null,
}

export interface ProtectionRequest {
    id: string,
    requesterAddress: string,
    decisions: LegalOfficerDecision[],
    userIdentity: Identity,
    userPostalAddress: PostalAddress,
    createdOn: string
}

export const LIGHT_MODE: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: rgbaToHex('#3b6cf4', 0.05),
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: rgbaToHex('#3b6cf4', 0.15),
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: rgbaToHex('#3b6cf4', 0.05),
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    }
};
