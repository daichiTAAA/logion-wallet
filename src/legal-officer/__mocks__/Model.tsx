import {
    fetchRecoveryInfo,
} from './ModelMock';

export {
    fetchRecoveryInfo,
};
