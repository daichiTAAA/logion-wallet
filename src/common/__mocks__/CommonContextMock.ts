import { TEST_WALLET_USER } from '../../wallet-user/TestData';
import { COLOR_THEME } from '../TestData';
import Addresses, { AccountAddress } from '../types/Addresses';

export let selectAddress = jest.fn();

export let currentAddress = TEST_WALLET_USER;

export function setCurrentAddress(address: string) {
    currentAddress = address;
}

const DEFAULT_ACCOUNT: AccountAddress = {
    name: "name",
    address: TEST_WALLET_USER,
    isLegalOfficer: false,
}

export let addresses: Addresses = {
    currentAddress: DEFAULT_ACCOUNT,
    addresses: [ DEFAULT_ACCOUNT ]
}

export let balances: any = null;

export let transactions: any = null;

export function useCommonContext() {
    return {
        currentAddress,
        selectAddress,
        addresses,
        balances,
        transactions,
        colorTheme: COLOR_THEME,
    };
}

export function setAddresses(value: Addresses) {
    addresses = value;
}

export function setBalances(value: any) {
    balances = value;
}

export function setTransactions(value: any) {
    transactions = value;
}
