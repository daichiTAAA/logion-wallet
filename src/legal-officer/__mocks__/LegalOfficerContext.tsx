export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = () => {

};

export let pendingProtectionRequests: any[] | null = null;

export let acceptedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export function useLegalOfficerContext() {
    return {
        pendingTokenizationRequests,
        rejectRequest,
        rejectedTokenizationRequests,
        acceptedTokenizationRequests,
        refreshRequests,
        pendingProtectionRequests,
        acceptedProtectionRequests,
        protectionRequestsHistory,
    };
}

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setAcceptedProtectionRequests(requests: any[]) {
    acceptedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}
