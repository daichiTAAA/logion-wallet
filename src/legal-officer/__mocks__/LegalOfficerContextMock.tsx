export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = jest.fn();

export let pendingProtectionRequests: any[] | null = null;

export let activatedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export let recoveryRequestsHistory: any[] | null = null;

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setActivatedProtectionRequests(requests: any[]) {
    activatedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}

export function setRecoveryRequestsHistory(requests: any[]) {
    recoveryRequestsHistory = requests;
}
