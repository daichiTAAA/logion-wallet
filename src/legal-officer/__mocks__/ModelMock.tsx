export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

export let rejectRequest = jest.fn();

export let acceptRequest = jest.fn();

export function setAcceptRequest(mockFn: any) {
    acceptRequest = mockFn;
}

export let setAssetDescription = jest.fn();

export let rejectProtectionRequest = jest.fn();

export function setRejectProtectionRequest(mockFn: any) {
    rejectProtectionRequest = mockFn;
}

export let acceptProtectionRequest = jest.fn();

export function setAcceptProtectionRequest(mockFn: any) {
    acceptProtectionRequest = mockFn;
}

export function decision(legalOfficerAddress: string, decisions: any[]) {
    return decisions[0];
}

export let fetchRecoveryInfo = jest.fn();

export function setFetchRecoveryInfo(mockFn: any) {
    fetchRecoveryInfo = mockFn;
}
