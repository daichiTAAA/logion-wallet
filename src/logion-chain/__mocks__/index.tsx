import { Extension } from '../Keys';

const LogionChainContextProvider = (props: any) => null;

let context = {
    apiState: 'CONNECT_INIT',
    injectedAccounts: []
};

export function setContextMock(value: any) {
    context = value;
}

export function useLogionChain() {
    return context;
}

let extensionAvailable = false;

export function setExtensionAvailable(value: boolean) {
    extensionAvailable = value;
}

export function isExtensionAvailable() {
    return extensionAvailable;
}

let recommendedExtensionValue: Extension | null = null;

export function setRecommendedExtension(value: Extension | null) {
    recommendedExtensionValue = value;
}

export function recommendedExtension(): Extension | null {
    return recommendedExtensionValue;
}

export {
    LogionChainContextProvider,
};
