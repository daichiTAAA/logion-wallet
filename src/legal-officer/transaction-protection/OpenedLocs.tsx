import React from 'react';

import { useCommonContext } from '../../common/CommonContext';
import Table, { Cell, EmptyTableMessage, DateTimeCell, ActionCell } from '../../common/Table';
import LocStatusCell from '../../common/LocStatusCell';
import LocIdCell from '../../common/LocIdCell';
import UserIdentityNameCell from '../../common/UserIdentityNameCell';
import Button from "../../common/Button";
import { useNavigate } from "react-router-dom";
import { locDetailsPath } from "../LegalOfficerPaths";
import ButtonGroup from "../../common/ButtonGroup";
import { LocType } from '../../logion-chain/Types';

export interface Props {
    locType: LocType;
}

export default function OpenedLocs(props: Props) {
    const { openedLocRequests, openedIdentityLocs } = useCommonContext();
    const navigate = useNavigate();

    if (openedLocRequests === null || openedIdentityLocs === null) {
        return null;
    }

    const requests = props.locType === 'Transaction' ? openedLocRequests : openedIdentityLocs;

    return (
        <Table
            columns={[
                {
                    "header": "Requester",
                    render: request => <UserIdentityNameCell userIdentity={ request.userIdentity }/>,
                    align: 'left',
                },
                {
                    "header": "Description",
                    render: request => <Cell content={ request.description } />,
                    align: 'left',
                },
                {
                    header: "Status",
                    render: request => <LocStatusCell status={ request.status }/>,
                    width: "140px",
                },
                {
                    header: "LOC ID",
                    render: request => <LocIdCell status={ request.status } id={ request.id } />,
                    align: "left",
                },
                {
                    header: "Creation date",
                    render: request => <DateTimeCell dateTime={ request.createdOn || null } />,
                    width: '200px',
                    align: 'center',
                },
                {
                    header: "Action",
                    render: request =>
                        <ActionCell>
                            <ButtonGroup>
                                <Button onClick={ () => navigate(locDetailsPath(request.id)) }>Manage LOC</Button>
                            </ButtonGroup>
                        </ActionCell>
                    ,
                    width: '200px',
                    align: 'center',
                }
            ] }
            data={ requests }
            renderEmpty={ () => <EmptyTableMessage>No LOCs</EmptyTableMessage> }
        />
    );
}
