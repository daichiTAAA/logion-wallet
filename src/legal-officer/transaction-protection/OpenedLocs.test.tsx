jest.mock('../../common/CommonContext');

import { IdentityLocType } from "../../logion-chain/Types";
import { shallowRender, render } from '../../tests';
import OpenedLocs from './OpenedLocs';
import { setOpenedLocRequests, setOpenedIdentityLocsByType } from '../../common/__mocks__/CommonContextMock';

test("Renders null with no data", () => {
    const tree = shallowRender(<OpenedLocs locType="Transaction" />);
    expect(tree).toMatchSnapshot();
});

test("Renders pending requests", () => {
    setOpenedLocRequests([
        {
            request: {
                id: "1",
                ownerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
                requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
                description: "LOC description",
                status: "OPEN"
            }
        }
    ]);
    const tree = shallowRender(<OpenedLocs locType="Transaction" />);
    expect(tree).toMatchSnapshot();
});

test("Renders Open Polkadot Identity LOCs", () => {
    const locs: Record<IdentityLocType, any[]> = {
        "Polkadot": [ { request: request("John", "Doe") } ],
        "Logion": []
    }
    setOpenedIdentityLocsByType(locs);
    const tree = render(<OpenedLocs locType="Identity" identityLocType="Polkadot" />);
    expect(tree).toMatchSnapshot();
});

test("Renders Open Logion Identity LOCs", () => {
    const locs: Record<IdentityLocType, any[]> = {
        "Polkadot": [],
        "Logion": [ { request: request("Scott", "Tiger") } ]
    }
    setOpenedIdentityLocsByType(locs);
    const tree = render(<OpenedLocs locType="Identity" identityLocType="Logion" />);
    expect(tree).toMatchSnapshot();
});

function request(firstName: string, lastName: string) {
    return {
        id: "556f4128-4fc3-4fdc-a543-74e6230911c4",
        ownerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        description: "LOC description",
        status: "OPEN",
        userIdentity: {
            firstName,
            lastName,
        }
    }
}
