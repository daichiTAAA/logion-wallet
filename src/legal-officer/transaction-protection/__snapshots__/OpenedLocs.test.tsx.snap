// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders null with no data 1`] = `null`;

exports[`Renders pending requests 1`] = `
<Table
  columns={
    Array [
      Object {
        "align": "left",
        "header": "Requester",
        "render": [Function],
      },
      Object {
        "align": "left",
        "header": "Description",
        "render": [Function],
      },
      Object {
        "header": "Status",
        "render": [Function],
        "width": "140px",
      },
      Object {
        "align": "left",
        "header": "LOC ID",
        "render": [Function],
      },
      Object {
        "align": "center",
        "header": "Creation date",
        "render": [Function],
        "width": "200px",
      },
      Object {
        "align": "center",
        "header": "Action",
        "render": [Function],
        "width": "200px",
      },
    ]
  }
  renderEmpty={[Function]}
/>
`;
